### ISSUE TITLE
Regression in 4.23.1: Log table is created multiple times

### ISSUE THREAD

**Issue by mhalbritter** [liquibase/liquibase#4684]
### Search first

- [X] I searched and no similar issues were found

### Description

Hey, I'm Moritz from the Spring Boot team. While updating Liquibase to 4.23.1, one of our tests starts failing. The test is setting custom table names for the changelog and the lock table, something like this:

```java
SpringLiquibase liquibase = ...
liquibase.setDatabaseChangeLogTable("log");
liquibase.setDatabaseChangeLogLockTable("lock");
```

With 4.23.0, this works fine. With 4.23.1, it starts failing with:

```
Aug. 16, 2023 4:07:05 PM liquibase.database
INFORMATION: Set default schema name to PUBLIC
Aug. 16, 2023 4:07:05 PM liquibase.changelog
INFORMATION: Creating database history table with name: PUBLIC.log
Aug. 16, 2023 4:07:05 PM liquibase.changelog
INFORMATION: Reading from PUBLIC.log
Aug. 16, 2023 4:07:05 PM liquibase.lockservice
INFORMATION: Changelog-Protokoll erfolgreich gesperrt.
Aug. 16, 2023 4:07:05 PM liquibase.changelog
INFORMATION: Creating database history table with name: PUBLIC.log
Aug. 16, 2023 4:07:05 PM liquibase.command
INFORMATION: Update command encountered an exception.
Aug. 16, 2023 4:07:05 PM liquibase.lockservice
INFORMATION: Successfully released change log lock
Aug. 16, 2023 4:07:05 PM liquibase.command
INFORMATION: Command execution complete
16:07:05.516 [main] WARN org.springframework.context.annotation.AnnotationConfigApplicationContext -- Exception encountered during context initialization - cancelling refresh attempt: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'liquibase' defined in org.springframework.boot.autoconfigure.liquibase.LiquibaseAutoConfiguration$LiquibaseConfiguration: liquibase.exception.DatabaseException: liquibase.exception.DatabaseException: Tabelle "LOG" besteht bereits
Table "LOG" already exists; SQL statement:
CREATE TABLE PUBLIC.log (ID VARCHAR(255) NOT NULL, AUTHOR VARCHAR(255) NOT NULL, FILENAME VARCHAR(255) NOT NULL, DATEEXECUTED TIMESTAMP NOT NULL, ORDEREXECUTED INT NOT NULL, EXECTYPE VARCHAR(10) NOT NULL, MD5SUM VARCHAR(35), DESCRIPTION VARCHAR(255), COMMENTS VARCHAR(255), TAG VARCHAR(255), LIQUIBASE VARCHAR(20), CONTEXTS VARCHAR(255), LABELS VARCHAR(255), DEPLOYMENT_ID VARCHAR(10)) [42101-220] [Failed SQL: (42101) CREATE TABLE PUBLIC.log (ID VARCHAR(255) NOT NULL, AUTHOR VARCHAR(255) NOT NULL, FILENAME VARCHAR(255) NOT NULL, DATEEXECUTED TIMESTAMP NOT NULL, ORDEREXECUTED INT NOT NULL, EXECTYPE VARCHAR(10) NOT NULL, MD5SUM VARCHAR(35), DESCRIPTION VARCHAR(255), COMMENTS VARCHAR(255), TAG VARCHAR(255), LIQUIBASE VARCHAR(20), CONTEXTS VARCHAR(255), LABELS VARCHAR(255), DEPLOYMENT_ID VARCHAR(10))]
```

It tries to create the log table two times. For reasons unknown to me, the tables are not found when running `liquibase.changelog.StandardChangeLogHistoryService#init` the 2nd time:

```
        Table changeLogTable = null;
        try {
            changeLogTable = SnapshotGeneratorFactory.getInstance().getDatabaseChangeLogTable(new SnapshotControl
                (database, false, Table.class, Column.class), database);
        } catch (LiquibaseException e) {
            throw new UnexpectedLiquibaseException(e);
        }
```

`changeLogTable` is null the 2nd time, too.

Workaround: Either quote the table names or use something like a `-` in it, so that Liquibase itself quotes it.

### Steps To Reproduce

Apply this patch to `main` of Spring boot and run the `org.springframework.boot.actuate.liquibase.LiquibaseEndpointTests#invokeWithCustomTables` test.

```
Index: spring-boot-project/spring-boot-dependencies/build.gradle
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/spring-boot-project/spring-boot-dependencies/build.gradle b/spring-boot-project/spring-boot-dependencies/build.gradle
--- a/spring-boot-project/spring-boot-dependencies/build.gradle	(revision 7d00ff4adf88164620377171346292ef0b014fa0)
+++ b/spring-boot-project/spring-boot-dependencies/build.gradle	(date 1692192619164)
@@ -810,7 +810,7 @@
 			]
 		}
 	}
-	library("Liquibase", "4.23.0") {
+	library("Liquibase", "4.23.1") {
 		group("org.liquibase") {
 			modules = [
 				"liquibase-cdi",
```

### Expected/Desired Behavior

It should work without quotes.

### Liquibase Version

4.23.1

### Database Vendor & Version

H2, but shouldn't matter

### Liquibase Integration

Spring Boot

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR (Thank you!)

**Comment by filipelautert** [comment]
Hello @mhalbritter - thanks for the report! 
After an initial investigation we narrowed down the problem to H2 database. There  are some workarounds as setting the table as all upper case, adding quotes, starting with a number or adding "-" to the table name. 
We will work to get a PR to fix it asap, thanks for the report!

**Comment by mhalbritter** [comment]
Thank you very much!

**Comment by stijn-vranckx** [comment]
The same thing occurs with a Postgres database but maybe for a different reason. 

In the validate method of DatabaseChangeLog you update the value of the ObjectQuotingStrategy. I would expect that at the end of the method you would put the current value back. By not doing this the tables are not found when running liquibase.changelog.StandardChangeLogHistoryService#init the 2nd time. This because the QUOTE_ALL_OBJECTS is used because of the LEGACY needed for Postgres.

**Comment by stijn-vranckx** [comment]
Should I log a separate issue for this problem?

**Comment by filipelautert** [comment]
Hello @stijn-vranckx - I believe this is a different issue as the one reported here happens because of changes introduced at H2Database class  . Could you please open a new issue for this one?

**Comment by izeye** [comment]
`LiquibaseEndpointTests.invokeWithCustomTables()` still seems to fail with 4.23.2 that contains 19b8710efa2a7033b89d98b22249d8c9a6378be4 as follows:

```
% git diff                                                                                     
diff --git a/spring-boot-project/spring-boot-dependencies/build.gradle b/spring-boot-project/spring-boot-dependencies/build.gradle
index e07a36ed83..df04772666 100644
--- a/spring-boot-project/spring-boot-dependencies/build.gradle
+++ b/spring-boot-project/spring-boot-dependencies/build.gradle
@@ -827,7 +827,7 @@ bom {
                        ]
                }
        }
-       library("Liquibase", "4.23.0") {
+       library("Liquibase", "4.23.2") {
                prohibit {
                        versionRange "4.23.1"
                        because "it contains a regression (https://github.com/liquibase/liquibase/issues/4684)"
% ./gradlew clean :spring-boot-project:spring-boot-actuator:test --tests LiquibaseEndpointTests

> Task :spring-boot-project:spring-boot-actuator:test

LiquibaseEndpointTests > invokeWithCustomTables() FAILED
    java.lang.IllegalStateException at LiquibaseEndpointTests.java:105
        Caused by: org.springframework.beans.factory.BeanCreationException at LiquibaseEndpointTests.java:104
            Caused by: liquibase.exception.CommandExecutionException at LiquibaseEndpointTests.java:104
                Caused by: liquibase.exception.DatabaseException at LiquibaseEndpointTests.java:104
                    Caused by: liquibase.exception.DatabaseException at LiquibaseEndpointTests.java:104
                        Caused by: org.h2.jdbc.JdbcSQLSyntaxErrorException at LiquibaseEndpointTests.java:104
...
```

**Comment by filipelautert** [comment]
Hello @izeye - this fix wasn't added on 4.23.2 - we had to revert a lot of stuff before release to make sure it was based solely on 4.23.1. 
But now it is going to be shipped with 4.24.0 that is probably going to be released this week.
